### PR TITLE
[Data] Change offsets to int64 and change to LargeList for ArrowTensorArray

### LINKED ISSUE #46434: [Data] ArrowVariableShapedTensorArray with LargeListArray
### Description

The current implementation only allows to create `ArrowVariableShapedTensorArray` objects with a maximum number of (2^31)-1 elements because it uses PyArrow's `ListArray` in `ray.air.util.tensor_extention.arrow` [L812](https://github.com/ray-project/ray/blob/b643ed8a0e906ebe19034e53fde8a215459022fc/python/ray/air/util/tensor_extensions/arrow.py#L812C1-L813C1) which uses 32-bit encoding for indexing. Thus, storing some types of data like long time-series which contain more elements than with 32-bit encoding causes overflow.

Providing the possibility to replace `ListArray` with Pyarrow [LargeListArray](https://arrow.apache.org/docs/python/generated/pyarrow.LargeListArray.html) would allow to store arrays with up to (2^63)-1 elements. (**Note**: this would also require to change the `OFFSET_DTYPE` in [L722](https://github.com/ray-project/ray/blob/b643ed8a0e906ebe19034e53fde8a215459022fc/python/ray/air/util/tensor_extensions/arrow.py#L722))

### Use case

The goal is to be able to store long time-series in arrow format (like long audios, or audios with high sample frequencies).

### PR DESCRIPTION
**Pull request by terraflops1048576**
Closes #46434

## Why are these changes needed?
Currently, the ArrowTensorArray and ArrowVariableShapedTensorArray types use int32 to encode the list offsets. This means that within a given PyArrow chunk in a column, the sum of the sizes of all of the tensors in that chunk must be less than 2^31; otherwise, depending on the overflow conditions, an error is thrown or the data is truncated. This usually doesn't manifest itself in Ray Data with the default settings, because it splits the blocks up to meet the target max block size (though this can be turned off!). However, it unavoidably shows up when one needs a large local shuffle buffer to feed into Ray Train.

This PR changes the offsets to be stored in 64-bit integers and updates the corresponding storage types of the TensorArrays.

As an example:

```
import ray.train
import ray.train.torch
import ray.data
import numpy as np

def f(batch):
   block1 = {"x": [np.ones((1000, 550), dtype=np.float16)] * 1000}
   return block1

dataset = ray.data.from_items([1, 2, 3, 4, 5]).map_batches(f, batch_size=None)
def train():
  data = ray.train.get_dataset_shard("train")
  for batch in data.iter_torch_batches(batch_size=100, local_shuffle_buffer_size=4000):
    pass

trainer = ray.train.torch.TorchTrainer(train_loop_per_worker=train, datasets={"train": dataset})
trainer.fit()
```

fails with
```
pyarrow.lib.ArrowInvalid: offset overflow while concatenating arrays
```
## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
Hi @terraflops1048576, thanks for fix! Can you also add a unit test?

**Comment by terraflops1048576** [comment]
I wasn't sure what unit test I should exactly add -- this PR is meant to change the list so its behavior still aligns with the existing unit tests, but works for large numbers of tensor elements. I added a unit test that checks that it actually works with the large numbers of tensors.

**Review by scottjlee — APPROVED** [review]
thanks for the fix! merging in master since it's been a few weeks, to ensure tests still pass

**Review by raulchen — APPROVED** [review]
Nice fix, thank you!

**Comment by terraflops1048576** [comment]
I think some of the test failures are unrelated, something to do with rllib. I see that one of the tests checks the size of the array; by my calculation, the array should use exactly 768 bytes to represent the tensors themselves, with 6 * 15 * 8 = 720 bytes for the values themselves, along with one 64-bit offset value per tensor. The rest is in the shape serialization, I believe. I updated the test to more accurately calculate the size, and hopefully it's within tolerance.

**Comment by can-anyscale** [comment]
Breaking many release tests: https://github.com/ray-project/ray/issues/46499, https://github.com/ray-project/ray/issues/46496, https://github.com/ray-project/ray/issues/46495. I'm reverting to unblock the release.
